Pre parse callback

phlptp opened the pull request:
This PR adds a pre parse callback and clarifies the callback timing of subcommands and option groups.  

It will also address issue #166.   The basic concept is to allow each app/subcommand/option_group to have two callbacks,  one would execute after the first argument  of the subcommand/option group is processed.  This allows an ability to effect the parsing dynamically, based on the number of arguments passed like Issue #166, or to dynamically enable or disable groups of options based on which options or subcommands were passed, or whatever other program state is appropriate to an application.    The other callback executes after parsing is complete.    With `immediate_callback` this is done immediately following the parse.    So callbacks from multiple subcommands can be executed in order and executed multiple times.    If that flag is not set all callbacks are executed on completion of all parsing in order of creation.  

To facilitate adaptive enabling or disabling option_groups some flags were created `enabled_by_default` and `disabled_by_default` that reset the disabled_ flag to a defined value on start of parsing.  

In the process of this a few issues with subcommand termination were fixed, and the readme updated with a new section on callbacks.  Also a subcommand_terminator was added (`++`)  to enable explicitly specifying that a subcommand is finished.   

Also added two new examples `shapes` with repeated subcommands and callbacks
and positional_arity which is Issue #166 coded up.
TODO: 
- [x] add missing tests for code coverage
- [x] add some helper functions to simplify triggered option groups

I am not sure I like all the names of the flags, so suggestions are welcome.

phlptp committed:
add some unit tests for the fallthrough_parent command

rework return values from _parse_* function to return true if the value was processed false otherwise, this simplified the logic and got rid of the pulling and clearing of the missing fields from option groups.

add TriggerOff and TriggerOn helper functions and some tests for them

add shapes example of multiple callbacks in order.

allow specification of callbacks that get executed immediately on completion of parsing of subcommand

add tests for enabled/disabled by default

add _get_fallthrough_parent.  To get the most appropriate parent to fallthrough to

add enabled and disabled by default functions

add positional_arity example

Add a pre_parse_callback_ for apps.  The Pre parse callback takes an argument for the number of remaining arguments left to process, and will execute prior to parsing for subcommands, and after the first option parse for option_groups.

phlptp committed:
Add excludes test on option_groups

phlptp committed:
make immediate_callback inheritable, add tests for that and for valid strings, some cleanup in the README.md

Apply suggestions from code review

Co-Authored-By: phlptp <top1@llnl.gov>
allow callbacks for option_groups, and allow some other characters as flags

henryiii commented on README.md:
````diff
@@ -235,7 +236,7 @@ Option_group *app.add_option_group(name,description); // 🚧
 ```
 
-An option name must start with a alphabetic character, underscore, or a number. For long options, after the first character '.', and '-' are also valid. For the `add_flag*` functions '{' has special meaning. Names are given as a comma separated string, with the dash or dashes. An option or flag can have as many names as you want, and afterward, using `count`, you can use any of the names, with da ...
+An option name must start with a alphabetic character, underscore, a number🚧, or one 🚧 of '$', '?', '#', '@', '%'. For long options, after the first character '.', and '-' are also valid characters. For the `add_flag*` functions '{' has special meaning. Names are given as a comma separated string, with the dash or dashes. An option or flag can have as many names as you want, and afterward, using  ...
````
Why $? That's hard to add in most bash-like shells.

henryiii replied:
It might be okay to support $, it just seems a bit odd. I know that `-?` can be useful.

phlptp replied:
/? is pretty standard for help in windows applications, and as I am converting other applications most of the time I just support help with -? and that failed the first time I tried it. So that is the most useful one.  The others I used -@ once as a shortcut for --address.  I guess I can see `$` being potentially problematic in bash.  If you wanted to remove all but '?'  that is probably fine.

henryiii replied:
I'd like to drop `$` to avoid issues with bash shells (adding a `--$this` option would not behave as expected), similarly `#` and `%` (comments in bash, Variables in windows) `?` is safe, as is `@` I think. However, the code could support `$#%` still; I just don't think they should be advertised. The only issue is if you want to reserve these as shortcuts, like the `!{}` chars.

henryiii commented on README.md:
```diff
@@ -505,9 +508,48 @@ There are several options that are supported on the main app and subcommands and
+
+#### Callbacks 
+A subcommand has two optional callbacks that are executed at different stages of processing.  The `preparse_callback` 🚧 is executed once after the first argument of a subcommand or application is processed and gives an argument for the number of remaining arguments to process.  For the main app the first argument is considered the program name,  for subcommands the first argument is the subcomman ...
+The second callback is executed after parsing.  Depending on the status of the `immediate_callback` flag 🚧.  This is either immediately after the parsing of the subcommand.  Or if the flag is false, once after parsing of all arguments.  If the immediate_callback is set then the callback can be executed multiple times.   If the main app or subcommand has a config file, no data from the config file ...
```
```suggestion
The second callback is executed after parsing.  The behavior depends on the status of the `immediate_callback` flag 🚧. If true, this runs immediately after the parsing of the subcommand.  Or if the flag is false, once after parsing of all arguments.  If the `immediate_callback` is set then the callback can be executed multiple times if the subcommand list given multiple times.  If the main app or subcommand has a config file, no data from the config file will be reflected in immediate_callback. `immediate_callback()` has no effect on the main app, though it can be inherited.  For option_groups `immediate_callback` causes the callback to be run prior to other option groups and options in the main app, effectively giving the options in the group priority.  
```

henryiii commented on README.md:
````diff
@@ -505,9 +508,48 @@ There are several options that are supported on the main app and subcommands and
+
+... A bunch of other options
+
+```
````
Remove ending blank line inside the code fence, add blank lines around all code fences (some markdown parsers will not parse correctly if a code fence is not surrounded by whitespace).

henryiii commented on README.md:
```diff
@@ -505,9 +508,48 @@ There are several options that are supported on the main app and subcommands and
+* c2 will be called once after processing all arguments 
+* ac will be called after completing the parse and all lower level callbacks have been executed
+
+A subcommand is considered terminated when one of the following conditions are met.
```
Add a blank line before starting (or ending) a list. Some markdown parsers need space.

henryiii commented on include/CLI/App.hpp:
```diff
@@ -90,6 +90,16 @@ class App {
+    bool pre_parse_called_{false};
+
+    /// Flag indicating that the callback for the subcommand should be executed immediately on parse completion which is
+    /// before help or ini files are processed.
```
This should be inheritable.

henryiii commented on include/CLI/App.hpp:
```diff
@@ -292,6 +314,25 @@ class App {
+        return this;
+    }
+
+    /// Set the subcommand to be enabled by default, so on clear(), at the start of each parse it is enabled(not
```
```suggestion
    /// Set the subcommand to be enabled by default, so on clear(), at the start of each parse it is enabled (not
```

henryiii commented on include/CLI/App.hpp:
```diff
@@ -1720,44 +1778,58 @@ class App {
             return detail::Classifier::SHORT;
         if((allow_windows_style_options_) && (detail::split_windows_style(current, dummy1, dummy2)))
             return detail::Classifier::WINDOWS;
+        if((current == "++") && (!name_.empty()))
```
```suggestion
        if((current == "++") && !name_.empty())
```

(Make it slightly easier to parse by eye)

henryiii commented on include/CLI/App.hpp:
```diff
@@ -1811,20 +1883,34 @@ class App {
 
         for(App_p &sub : subcommands_) {
-            sub->_process_env();
+            if((sub->get_name().empty()) || (!sub->immediate_callback_))
```
```suggestion
            if(sub->get_name().empty() || !sub->immediate_callback_)
```

henryiii commented on include/CLI/App.hpp:
```diff
@@ -1811,20 +1883,34 @@ class App {
+
+        for(App_p &sub : subcommands_) {
+            // process the priority option_groups first
+            if((sub->get_name().empty()) && (sub->immediate_callback_)) {
```
```suggestion
            if(sub->get_name().empty() && sub->immediate_callback_) {
```

henryiii commented on include/CLI/App.hpp:
```diff
@@ -1952,7 +2038,10 @@ class App {
+                sub->_process_requirements();
+            }
+
             if((sub->required_) && (sub->count_all() == 0)) {
```
(I think it's better not to add () if they are not needed, but it's done on this way on the line below that was not changed by this PR, so it's okay to leave them for now if you prefer. I might try to clean them up one day.

phlptp replied:
I cleaned up a few of them as I was looking through other suggestions, I am sure there are more.

henryiii commented on include/CLI/App.hpp:
```diff
@@ -2116,52 +2234,59 @@ class App {
             }
         }
-
+        /// let the parent deal with it if possible
```
Stylistic note: Comments inline are two `//`, three slashes indicated that Doxygen picks up the comment for the API docs, and it can't pick up comments in code.

henryiii commented on include/CLI/App.hpp:
```diff
@@ -2408,6 +2573,47 @@ class Option_group : public App {
         add_options(args...);
     }
 };
+/// Helper function to enable one option group/subcommand when another is used
```
It's hard to tell here, but these should be in namespace detail, correct? (Or a nameless namespace)

phlptp replied:
I wasn't entirely sure where to put these.  They are intended to used by users similarly to the Validators.  They could be member functions of app.  They could be classes on their own.    I wanted the functionality but I am not sure of the name or location.  Basically a way to turn on/off option groups in a sequence during a parse.

henryiii replied:
Okay, they have a slightly different use than I thought.

phlptp commented on include/CLI/App.hpp:
```diff
@@ -1720,44 +1778,58 @@ class App {
         }
     }
-    /// Internal function to run (App) callback, top down
+    /// Internal function to run (App) callback, bottom up
```
This is a change in behavior,  the order of callbacks between the main app and child subcommands.  Previously it was a top down approach.  Now it is a bottom up approach.  There wasn't any tests that relied on this behavior but it is a change.   I would argue it makes more sense as a bottom up approach, to know that the main app callback only executes after everything else has completed.  But being there was a comment on the effect I wasn't entirely sure if there was a reason behind it or something that depended on that behavior.    I would say in most applications it would make no difference, but this is a change in behavior.

henryiii replied:
Would a user be able to do something like:

```bash
git -C dir add -u 
```

That is, add would need to change to the `dir` directory. It's not too important, for this example or the existing code, since you don't have to run this in callbacks; you can just do a complete parse then inspect the results, which I assume is a bit more common.

henryiii replied:
PS: I'm okay with the change, if I didn't say that clearly at the end (I didn't)

phlptp committed:
change add_option_function to take a std::function with void return type and update some corresponding test cases and documentation

Allow option groups to use ignore_case ignore_underscore for inheritance

Allow option groups to specify allow_extras even if the parent app doesn't in which case extra options flow down into the option_group.

phlptp committed:
add docs for remove_subcommand and add_subcommand in option_group

add some test of the remove_excludes functions

add test for Issue #256

add remove_subcommand fail test

add remove_subcommand function and add_subcommand to option_group and some tests associated with them.
